### PR TITLE
Montage fix

### PR DESCRIPTION
**Pull request by wmvanvliet**
This one looks better. It makes some small fixes to your PR.

### CODE REVIEW

**Comment by dengemann** [comment]
Ah ok:) we leave the git-fu excercise for your next PR ;-) We normally try to avoid closing PRs for cleanup reasons (otherwise boom).

File: mne/montages/montage.py [hunk 1]
```diff
@@ -77,23 +77,25 @@ def read_montage(kind, names=None, path=None, scale=True):
...
+        montages = [op.splitext(f) for f in os.listdir(path)]
+        montages = [m for m in montages if m[1] in supported and kind == m[0]]
+        if len(montages) != 1:
             raise ValueError('Could not find the montage. Please provide the'
                              'full path')
-        kind = kind[0]
+        kind, ext = montages[0]
+        fname = op.join(path, kind + ext)
+    else:
+        kind, ext = op.splitext(kind)
```
Comment (dengemann) [inline mne/montages/montage.py:88]: nice!

File: mne/viz/montage.py [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+    fig : isntance of mayavi.Scene
+        The malab scene object.
+    """
+    fig = mlab.figure(bgcolor=(0.0, 0.0, 0.0), size=(600, 600))
+    pos = montage.pos
+    mlab.points3d(pos[:, 0], pos[:, 1], pos[:, 2],
+                  color=(1.0, 1.0, 1.0), scale_factor=1.5)
+    mlab.text(0.01, 0.01, montage.kind, width=0.4)
+    mlab.view(0, 0)
+    return fig
```
Comment (dengemann) [inline mne/viz/montage.py:31]: I assume this is the old function in its new correct place?

**Comment by wmvanvliet** [comment]
Yeah sorry about that. Rebasing proved to be a big pain in this case (somehow it flagged all files as being in conflict). My git-fu is not that strong yet.

File: mne/viz/montage.py (hunk 2 above)
Comment (wmvanvliet) [inline mne/viz/montage.py:31]: yes, a cut/paste job

File: mne/viz/tests/test_montage.py [hunk 3]
```diff
@@ -0,0 +1,19 @@
+import numpy as np
+from mne.montages import read_montage
+
+lacks_mayavi = False
```
Comment (dengemann) [inline mne/viz/tests/test_montage.py]: I think by conventions we call bool flags like this `has_`. We can test this using git grep in the source of the repository.

git grep 'has_'
git grep 'lacks_'

But probably my bad ;-)

Comment (dengemann) [inline mne/viz/tests/test_montage.py]: edited.

**Comment by dengemann** [comment]
Apart from that one comment LGTM. I'm currently busy on another branch, I'll merge and test tomorrow. Thanks a lot!

File: mne/viz/montage.py (hunk 2 above)
Comment (dengemann) [inline mne/viz/montage.py:31]: just wanted to makes sure I miss specific changes.

**Comment by dengemann** [comment]
> Yeah sorry about that.

Don't worry, git-fu is growing by doing PRs ;-)

**Comment by dengemann** [comment]
@wmvanvliet out of curiosity, how do you use the montage so far? Does it work for you?

File: mne/montages/montage.py [hunk 4]
```diff
@@ -120,12 +121,11 @@ def read_montage(kind, names=None, path=None, scale=True):
         z = 85. * np.cos(np.deg2rad(theta))
         pos = np.c_[x, y, z]
         names_ = data['f0']
```
Comment (dengemann) [inline mne/montages/montage.py:123]: @wmvanvliet did you have a chance to play with this montage? I think here was an issue with the orientation, I might be wrong.

File: mne/viz/tests/test_montage.py (hunk 3 above)
Comment (wmvanvliet) [inline mne/viz/tests/test_montage.py]: actually the rest of the code in `utils.py` uses `skip` as variable name.

File: mne/viz/montage.py (hunk 2 above)
Comment (wmvanvliet) [inline mne/viz/montage.py:31]: small typo in the docs there

File: mne/montages/montage.py (hunk 4 above)
Comment (wmvanvliet) [inline mne/montages/montage.py:123]: I've just visualized the one montage. Now that this code merges cleanly with the master branch, I can look into it a bit more.

File: mne/montages/montage.py [hunk 5]
```diff
@@ -164,7 +164,12 @@ def apply_montage(info, montage):
     if not _contains_ch_type(info, 'eeg'):
         raise ValueError('No eeg channels found')
     for pos, name in zip(montage.pos, montage.names):
-        ch_idx = info['ch_names'].index(name)
```
Comment (wmvanvliet) [inline mne/montages/montage.py:167]: I strongly feel that `apply_montage` should not fail if a sensor is defined in the montage file but not present in the `info` structure; such sensors should just be skipped. This way, the standard 10_5 montage can for example be applied to many EEG sets that follow the default names.

Comment (agramfort) [inline mne/montages/montage.py:167]: Fine with me

Comment (dengemann) [inline mne/montages/montage.py:167]: I would at least issue an error if none of the montage names is present in the info. Herr I would not use a try / except. Just continue if name is not in info['ch_names'].

**Comment by wmvanvliet** [comment]
I'm trying to load a BDF file I've recorded. It doesn't store any electrode positions, but follows the 10-20 naming convention, so it can be a good use case for the montage functionality. Some problems arise:
1. `read_raw_edf` should use the new montage loading functionality instead of requiring a hpts file that specified electrode locations.
2. support for hpts files should be added to `montage.py` to replace the existing functionality in `read_raw_edf` and keep things backwards compatible. The code to do this could be lifted from this function.
3. `fit_sphere_to_headshape` requires three cardinal landmarks: nasion, left auricular, right auricular. Without these, functions like `make_eeg_layout` will not work.

**Comment by dengemann** [comment]
> support for hpts files should be added to montage.py to replace the existing functionality in > read_raw_edf and keep things backwards compatible. The code to do this could be lifted from this function.

sounds like that will be another PR ...

> fit_sphere_to_headshape requires three cardinal landmarks: nasion, left auricular, right auricular. Without these, functions like make_eeg_layout will not work.

I think I once replaced the internals, don't remember where. I'm pretty sure you don't need fiducials.... Need to recap ...

**Comment by dengemann** [comment]
@wmvanvliet 

I think we need to add the data directory somewhere here to make Travis happy https://github.com/mne-tools/mne-python/blob/master/.travis.yml#L86

**Comment by dengemann** [comment]
@wmvanvliet see the environment definitions here for the tests? https://github.com/mne-tools/mne-python/blob/master/.travis.yml#L17

Some of them don't mix well with mayavi. I think we should skip the viz tests for all envs that don't install mayavi. See failing tests.

**Comment by wmvanvliet** [comment]
Sorry, I don't follow. What data directory? In `viz/tests/test_montage.py`, the test is skipped if mayavi is not installed.

**Comment by wmvanvliet** [comment]
Hrm, what does the `scale` flag do?
https://github.com/wmvanvliet/mne-python/blob/montage-fix/mne/montages/montage.py#L56

**Comment by agramfort** [comment]
just add mne/montages to setup.py

can you type the code the way you would like to type it? from raw data to add montage to a plot using topomap?

**Comment by wmvanvliet** [comment]
```
import mne

# Load BDF file, specifying standard 10-20 naming convention for electrode
# positions
raw = mne.io.read_raw_edf('semantic_priming_experiment.bdf', n_eeg=32, stim_channel=-1,
                          preload=True, hpts='standard_1020')

# Set bads and reference
raw.info['bads'] = ['T7']
raw, ref = mne.io.set_eeg_reference(raw, ['EXG5', 'EXG6'])

# Bandpass filter
raw.filter(0.5, 15, l_trans_bandwidth=0.2, method='iir')

# Extract epochs for two classes 'related' and 'unrelated'
events = mne.find_events(raw)
picks = mne.pick_types(raw.info, meg=False, eeg=True, exclude='bads')
event_id = dict(related=2, unrelated=3)
epochs = mne.Epochs(raw, events, event_id, tmin=-0.1, tmax=1.0, picks=picks,
                    baseline=(None, 0), reject=dict(eeg=100e-6),
                    preload=True)

# Create ERPs for both classes
related = epochs['related'].average()
unrelated = epochs['unrelated'].average()

# Plot the ERPs
l = mne.layouts.make_eeg_layout(epochs.info)
mne.viz.plot_topo([related, unrelated], l)
```

**Comment by agramfort** [comment]
so montage only gives you channel names?

the first think that comes to mind is that we should have a method of BaseRaw as other data formats will need montage.

raw.set_montage('standard_1020')

or maybe you have a better name in mind.

**Comment by wmvanvliet** [comment]
in the example above, the montage gives you electrode positions, based on the channel names defined in the BDF file.

the `montages.apply_montage` function applies the montage to any raw object. We could add it as a method of BaseRaw if we want.

Currently, the `read_raw_edf` function takes a `hpts` parameter to specify electrode positions. The new montage facility should deprecate this, as it's easier and more flexible. We could remove the `hpts` flag and require the user to manually call `apply_montage`. In the example I opted for backwards compatibility and specified the montage name as the `hpts` parameter.

**Comment by agramfort** [comment]
> in the example above, the montage gives you electrode positions, based on the channel names defined in the BDF file.
> 
> the montages.apply_montage function applies the montage to any raw object. We could add it as a method of BaseRaw if we want.

any raw object or any container that has an info attribute?

> Currently, the read_raw_edf function takes a hpts parameter to specify electrode positions. The new montage facility should deprecate this, as it's easier and more flexible. We could remove the hpts flag and require the user to manually call apply_montage. In the example I opted for backwards compatibility and specified the montage name as the hpts parameter.

is the semantic of hpts file and the montage really equivalent?
anyone? do most people have hpts files?

**Comment by dengemann** [comment]
> raw.set_montage('standard_1020')

or `raw.add_montage` as raw.add_proj

I like this idea.

> any raw object or any container that has an info attribute?

actually yes. Would add it to some mixin. Maybe the channels Mixin.

File: mne/utils.py [hunk 6]
```diff
@@ -722,24 +722,30 @@ def dec(*args, **kwargs):
     return dec
 
 
-def requires_mayavi():
+def requires_mayavi(function):
```
Comment (dengemann) [inline mne/utils.py:725]: @wmvanvliet I think this change creates trouble

**Comment by wmvanvliet** [comment]
> Is the semantic of hpts file and the montage really equivalent?

When loading EDF/BDF files, it is. The function expects a single point for each sensor, plus fiducials. Many EEG labs do not have any device to produce these points and would instead rely on approximate positions provided by the manufacturer of the electrode cap. So many do not have .hpts files.

**Comment by dengemann** [comment]
@wmvanvliet I'm a bit swamped at the moment. Feel free to move on. I'll be mergin and finalizing in the next days, probably rather next week. Please feel free to add your name to the files you've edited.

**Comment by dengemann** [comment]
Thanks @wmvanvliet ! I'm now finishing this.

**Comment by wmvanvliet** [comment]
hope it gets merges soon, this is a very useful feature when dealing with EEG data.

**Comment by dengemann** [comment]
We have a blocker. No merge unless the rotation issue is solved ... :(

**Pull request merged**